lib/runtime-core/src/jit_debug.rs

```diff
@@ -0,0 +1,195 @@
+//! Code for interacting with the
+//! [GDB JIT inteface](https://sourceware.org/gdb/current/onlinedocs/gdb.html#JIT-Interface).
+
+use std::ptr;
+use std::sync::Arc;
+
+/// Entrypoint that the debugger will use to trigger a read from the
+/// [`__jit_debug_descriptor`] global variable.
+///
+/// The debugger will wait for this function to be called and then take
+/// control to read the data we prepared.
+// Implementation of this function is derived from wasmtime and is licensed under
+// the Apache 2.0 license.  See ATTRIBUTIONS.md for full license and more
+// information.
+#[no_mangle]
+#[inline(never)]
+extern "C" fn __jit_debug_register_code() {
+    // This code exists to prevent optimization of this function so that the
+    // GDB JIT interface behaves as expected
+    let x = 3;
+    unsafe {
+        std::ptr::read_volatile(&x);
+    }
+}
+
+/// The operation that the debugger should perform with the entry that we gave it.
+#[allow(non_camel_case_types)]
+#[derive(Debug)]
+#[repr(u32)]
+pub(crate) enum JITAction {
+    /// Do nothing.
+    JIT_NOACTION = 0,
+    /// Register the given code.
+    JIT_REGISTER_FN = 1,
+    /// Unregister the given code.
+    JIT_UNREGISTER_FN = 2,
+}
+
+/// Node of the doubly linked list that the GDB JIT interface reads from.
+#[no_mangle]
+#[repr(C)]
+pub(crate) struct JITCodeEntry {
+    /// Next entry in the linked list.
+    next: *mut JITCodeEntry,
+    /// Previous entry in the linked list.
+    prev: *mut JITCodeEntry,
+    /// Pointer to the data we want the debugger to read.
+    symfile_addr: *mut u8,
+    /// The amount of data at the `symfile_addr` pointer.
+    symfile_size: u64,
+}
+
+impl Default for JITCodeEntry {
+    fn default() -> Self {
+        Self {
+            next: ptr::null_mut(),
+            prev: ptr::null_mut(),
+            symfile_addr: ptr::null_mut(),
+            symfile_size: 0,
+        }
+    }
+}
+
+/// Head node of the doubly linked list that the GDB JIT interface expects.
+#[no_mangle]
+#[repr(C)]
+pub(crate) struct JitDebugDescriptor {
```

Comment:
Are we sure we want `JitDebugDescriptor` to be public in the crate? I don't see any reason. Am I missing something?

```suggestion
struct JitDebugDescriptor {
```